x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,619 @@
+	iterator := sdk.KVStorePrefixIterator(store, iteratorPrefix)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		_, vscID, err := types.ParseChainIdAndVscIdKey(types.ConsumerValidatorsByVscIDBytePrefix, iterator.Key())
+		if err != nil {
+			panic(err)
+		}
+		var consumerAddrsToPrune types.AddressList
+		err = consumerAddrsToPrune.Unmarshal(iterator.Value())
+		if err != nil {
+			panic(err)
+		}
+		stop := cb(vscID, consumerAddrsToPrune.Addresses)
+		if stop {
+			break
+		}
+	}
+}
+
+func (k Keeper) IterateAllConsumerAddrsToPrune(
+	ctx sdk.Context,
+	cb func(chainID string, vscID uint64, consumerAddrsToPrune [][]byte) (stop bool),
+) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{types.ConsumerValidatorsByVscIDBytePrefix})
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		chainID, vscID, err := types.ParseChainIdAndVscIdKey(types.ConsumerValidatorsByVscIDBytePrefix, iterator.Key())
+		if err != nil {
+			panic(err)
+		}
+		var consumerAddrsToPrune types.AddressList
+		err = consumerAddrsToPrune.Unmarshal(iterator.Value())
+		if err != nil {
+			panic(err)
+		}
+		stop := cb(chainID, vscID, consumerAddrsToPrune.Addresses)
+		if stop {
+			break
+		}
+	}
+}
+
+// DeleteConsumerAddrsToPrune deletes the list of consumer addresses mapped to a given VSC ID
+func (k Keeper) DeleteConsumerAddrsToPrune(ctx sdk.Context, chainID string, vscID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ConsumerValidatorsByVscIDKey(chainID, vscID))
+}
+
+// AssignConsumerKey assigns the consumerKey to the validator with providerAddr
+// on the consumer chain with ID chainID
+func (k Keeper) AssignConsumerKey(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr sdk.ConsAddress,
+	consumerKey tmprotocrypto.PublicKey,
+) error {
+	consumerAddr := utils.TMCryptoPublicKeyToConsAddr(consumerKey)
+
+	// validator must already be registered
+	validator, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, providerAddr)
+	if !found {
+		return stakingtypes.ErrNoValidatorFound
+	}
+
+	if _, found := k.GetValidatorByConsumerAddr(ctx, chainID, consumerAddr); found {
+		// mapping already exists; return error
+		return sdkerrors.Wrapf(
+			types.ErrInvalidConsumerConsensusPubKey, "consumer key already exists",
+		)
+	}
+
+	// check whether the consumer chain is already registered,
+	// i.e., a client to the consumer was already created
+	if _, consumerRegistered := k.GetConsumerClientId(ctx, chainID); consumerRegistered {
+		// set the mapping from this validator's new consensus address on the consumer
+		// to its consensus address on the provider;
+		// otherwise, the mapping is added when the consumer is registered
+		// note: this state must be deleted through the pruning mechanism;
+		// see ConsumerValidatorsByVscID
+		k.SetValidatorByConsumerAddr(ctx, chainID, consumerAddr, providerAddr)
```

Comment:
Bug: this is only called if the consumer is registered so it is possible for many validators to assign the same consumer key before the chain is registered.

Reply:
Nice catch. See https://github.com/cosmos/interchain-security/pull/515/commits/86c597910b7ddd38a8ece96c7709813b3388591c